krm-functions/lib/fn-runtime/upstream_inventory.go

```diff
@@ -0,0 +1,126 @@
+package fnruntime
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	kptv1 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1"
+	"github.com/google/go-cmp/cmp"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type UpstreamInventory interface {
+	AddExistingCondition(*corev1.ObjectReference, *kptv1.Condition)
+	AddExistingResource(*corev1.ObjectReference, *fn.KubeObject)
+	AddNewResource(*corev1.ObjectReference, *fn.KubeObject)
+	Diff() (UpstreamInventoryDiff, error)
+}
+
+type UpstreamInventoryDiff struct {
+	DeleteObjs       []*Object
+	UpdateObjs       []*Object
+	CreateObjs       []*Object
+	DeleteConditions []*Object
+	CreateConditions []*Object
+	UpdateConditions []*Object
+}
+
+type Object struct {
+	Ref corev1.ObjectReference
+	Obj fn.KubeObject
+}
+
+func NewUpstreamInventory() UpstreamInventory {
+	return &upstreamInventory{
+		resources: map[corev1.ObjectReference]*upstreamInventoryCtx{},
+	}
+}
+
+type upstreamInventory struct {
+	m sync.RWMutex
+	resources map[corev1.ObjectReference]*upstreamInventoryCtx
+}
+
+type upstreamInventoryCtx struct {
+	existingCondition *kptv1.Condition
+	existingResource  *fn.KubeObject
+	newResource       *fn.KubeObject
+}
+
+func (r *upstreamInventory) AddExistingCondition(ref *corev1.ObjectReference, c *kptv1.Condition) {
+	r.m.Lock()
+	defer r.m.Unlock() 
+	if _, ok := r.resources[*ref]; !ok {
+		r.resources[*ref] = &upstreamInventoryCtx{}
+	}
+	r.resources[*ref].existingCondition = c
+
+}
+
+func (r *upstreamInventory) AddExistingResource(ref *corev1.ObjectReference, o *fn.KubeObject) {
+	r.m.Lock()
+	defer r.m.Unlock() 
+	if _, ok := r.resources[*ref]; !ok {
+		r.resources[*ref] = &upstreamInventoryCtx{}
+	}
+	r.resources[*ref].existingResource = o
+}
+
+func (r *upstreamInventory) AddNewResource(ref *corev1.ObjectReference, o *fn.KubeObject) {
+	r.m.Lock()
+	defer r.m.Unlock() 
+	if _, ok := r.resources[*ref]; !ok {
+		r.resources[*ref] = &upstreamInventoryCtx{}
+	}
+	r.resources[*ref].newResource = o
+}
+
+func (r *upstreamInventory) Diff() (UpstreamInventoryDiff, error) {
+	r.m.RLock()
+	defer r.m.RUnlock() 
+	diff := UpstreamInventoryDiff{
+		DeleteObjs:       []*Object{},
+		UpdateObjs:       []*Object{},
+		CreateObjs:       []*Object{},
+		DeleteConditions: []*Object{},
+		CreateConditions: []*Object{},
+	}
+
+	for ref, invCtx := range r.resources {
+		switch {
```

Comment:
Grouping switch statements is helpful. I guess some explanation would be helpful as well.

Reply:
added some comments that hopefully help to see what is happening